crates/sui-framework/sources/collectible/kiosk.move

```diff
@@ -0,0 +1,602 @@
+module sui::kiosk_tests {
+    use std::vector;
+    use sui::tx_context::{Self, TxContext};
+    use sui::transfer_policy::{Self as policy, TransferPolicy, TransferPolicyCap};
+    use sui::kiosk::{Self, Kiosk, KioskOwnerCap};
+    use sui::object::{Self, ID, UID};
+    use sui::sui::SUI;
+    use sui::package;
+    use sui::coin;
+
+    const AMT: u64 = 10_000;
+
+    struct Asset has key, store { id: UID }
+    struct OTW has drop {}
+
+    /// Prepare: accounts
+    /// Alice, Bob and my favorite guy - Carl
+    fun folks(): (address, address, address) { (@0xA11CE, @0xB0B, @0xCA51) }
+
+    /// Prepare: TransferPolicy<Asset>
+    fun get_policy(ctx: &mut TxContext): (TransferPolicy<Asset>, TransferPolicyCap<Asset>) {
+        let publisher = package::test_claim(OTW {}, ctx);
+        let (policy, cap) = policy::new(&publisher, ctx);
+        package::burn_publisher(publisher);
+        (policy, cap)
+    }
+
+    /// Prepare: Asset
+    fun get_asset(ctx: &mut TxContext): (Asset, ID) {
+        let uid = object::new(ctx);
+        let id = object::uid_to_inner(&uid);
+        (Asset { id: uid }, id)
+    }
+
+    #[test]
+    fun test_list_and_take() {
+        let ctx = &mut tx_context::dummy();
+        let (policy, policy_cap) = get_policy(ctx);
+        let (asset, item_id) = get_asset(ctx);
+        let (kiosk, owner_cap) = kiosk::new(ctx);
+
+        kiosk::place(&mut kiosk, &owner_cap, asset);
+        let asset = kiosk::take(&mut kiosk, &owner_cap, item_id);
+
+        return_kiosk(kiosk, owner_cap, ctx);
+        return_assets(vector[ asset ]);
+        return_policy(policy, policy_cap, ctx);
+    }
+
+    #[test]
+    fun test_purchase() {
+        let ctx = &mut tx_context::dummy();
+        let (policy, policy_cap) = get_policy(ctx);
+        let (asset, item_id) = get_asset(ctx);
+        let (kiosk, owner_cap) = kiosk::new(ctx);
+
+        kiosk::place_and_list(&mut kiosk, &owner_cap, asset, AMT);
+
+        let payment = coin::mint_for_testing<SUI>(AMT, ctx);
+        let (asset, request) = kiosk::purchase(&mut kiosk, item_id, payment, ctx);
+        policy::confirm_request(&mut policy, request);
+
+        return_kiosk(kiosk, owner_cap, ctx);
+        return_assets(vector[ asset ]);
+        return_policy(policy, policy_cap, ctx);
+    }
+
+    #[test]
+    #[expected_failure(abort_code = sui::kiosk::EIncorrectAmount)]
+    fun test_purchase_wrong_amount() {
+        let ctx = &mut tx_context::dummy();
+        let (policy, policy_cap) = get_policy(ctx);
+        let (asset, item_id) = get_asset(ctx);
+        let (kiosk, owner_cap) = kiosk::new(ctx);
+
+        kiosk::place_and_list(&mut kiosk, &owner_cap, asset, AMT);
+
+        let payment = coin::mint_for_testing<SUI>(AMT + 1, ctx);
+        let (asset, request) = kiosk::purchase(&mut kiosk, item_id, payment, ctx);
+        policy::confirm_request(&mut policy, request);
+        return_assets(vector[ asset ]);
+        return_policy(policy, policy_cap, ctx);
```

Comment:
Nit: you can get rid of all this because of the abort -- makes it clearer where the expected failure is supposed to come from.

Reply:
Yeee, nope, I can't, it emits a warning about unused or ignored value...

Reply:
Ah, that's probably for `asset` and `request`? I wonder if you could just not assign them, or assign them to underscored variables? (`_asset` and `_request`). Anyway, not a big deal.

Reply:
If you underscore something without drop, compiler yields a warning. That's quite lol, but trust me this is the way it works. I wouldn't do anything until it's necessary. 😆